### PR TITLE
[editor] Support migration for old versions of edits.

### PR DESCRIPTION
**Pull request by milchakov**
https://jira.mail.ru/browse/MAPSME-6593

### CODE REVIEW

File: editor/feature_matcher.cpp [hunk 1]
```diff
@@ -109,8 +110,9 @@ double MatchByGeometry(LGeometry const & lhs, RGeometry const & rhs)
 MultiPolygon TrianglesToPolygon(vector<m2::PointD> const & points)
 {
   size_t const kTriangleSize = 3;
-  CHECK_EQUAL(points.size() % kTriangleSize, 0, ());
-  CHECK(!points.empty(), ());
+  if (points.size() % kTriangleSize != 0 || points.empty())
+    MYTHROW(matcher::NotAPolygonException, ("Count of points must be multiple of", kTriangleSize));
```
Comment (mgsergio) [inline editor/feature_matcher.cpp]: Strictly speaking, zero is a multiple of anything but zero itself.

File: editor/feature_matcher.hpp [hunk 2]
```diff
@@ -5,16 +5,24 @@
 #include "geometry/mercator.hpp"
 #include "geometry/point2d.hpp"
 
+#include "base/exception.hpp"
+
 #include "std/vector.hpp"
```
Comment (mgsergio) [inline editor/feature_matcher.hpp]: #include <vector>

File: editor/feature_matcher.hpp [hunk 3]
```diff
@@ -5,16 +5,24 @@
...
+/// Throws NotAPolygon exception when |geometry| is not convertible to a polygon.
 pugi::xml_node GetBestOsmWayOrRelation(pugi::xml_document const & osmResponse,
                                        vector<m2::PointD> const & geometry);
 /// Returns value form [-1, 1]. Negative values are used as penalty, positive as score.
-/// |lhs| and |rhs| - triangulated polygons;
+/// |lhs| and |rhs| - triangulated polygons.
+/// Throws NotAPolygon exception when lhs or rhs is not convertible to a polygon.
 double ScoreTriangulatedGeometries(vector<m2::PointD> const & lhs, vector<m2::PointD> const & rhs);
+/// Deprecated, use ScoreTriangulatedGeometries instead.
+double ScoreTriangulatedGeometriesByPoints(vector<m2::PointD> const & lhs,
```
Comment (mgsergio) [inline editor/feature_matcher.hpp]: Why should we keep it?

File: indexer/edits_migration.cpp [hunk 4]
```diff
@@ -69,7 +68,19 @@ FeatureID MigrateWayorRelatonFeatureIndex(
...
-        auto const score = matcher::ScoreTriangulatedGeometries(geometry, ftGeometry);
+        double score = 0.0;
+        try
+        {
+          score = matcher::ScoreTriangulatedGeometries(geometry, ftGeometry);
+        }
+        catch (matcher::NotAPolygonException & ex)
+        {
+          // Support migration for old application versions.
+          // TODO(a): To remove it after some time.
```
Comment (mgsergio) [inline indexer/edits_migration.cpp]: What time exactly? If I work with this code after "some time" how can I tell it is the right time to remove it?

File: editor/feature_matcher.cpp [hunk 5]
```diff
@@ -345,4 +350,28 @@ double ScoreTriangulatedGeometries(vector<m2::PointD> const & lhs, vector<m2::Po
...
+  // used in set_itersection (cmp2) are compatible in that sence that
+  // cmp2(a, b) :- cmp1(a, b) and
+  // cmp1(a, b) :- cmp2(a, b) || a almost equal b.
+  // You can think of cmp2 as !(a >= b).
+  // But cmp2 is not transitive:
+  // i.e. !cmp(a, b) && !cmp(b, c) does NOT implies !cmp(a, c),
+  // |a, b| < eps, |b, c| < eps.
+  // This could lead to unexpected results in set_itersection (with greedy implementation),
+  // but we assume such situation is very unlikely.
+  auto const matched = set_intersection(begin(lhs), end(lhs),
```
Comment (mgsergio) [inline editor/feature_matcher.cpp]: This code looks very familiar to me. I see an addition of it, but where is the removing?

File: editor/feature_matcher.cpp (hunk 1 above)
Comment (milchakov) [reply editor/feature_matcher.cpp]: убрал || points.empty(), т.к. такой кейс должен обрабатываться как пустой полигон.

File: editor/feature_matcher.cpp (hunk 5 above)
Comment (milchakov) [reply editor/feature_matcher.cpp]: обсудили голосом.

File: editor/feature_matcher.hpp (hunk 2 above)
Comment (milchakov) [reply editor/feature_matcher.hpp]: готово

File: indexer/edits_migration.cpp (hunk 4 above)
Comment (milchakov) [reply indexer/edits_migration.cpp]: заменил на
To remove it when version 8.0.x will be not used.

File: editor/feature_matcher.hpp (hunk 3 above)
Comment (milchakov) [reply editor/feature_matcher.hpp]: /// Deprecated, use ScoreTriangulatedGeometries instead. In old versions duplicated points were removed
/// from list of points, but it not possible to make polygon by using this list. We need to keep
/// this method to support migration from old application vesions (8.0.x and below).

**Comment by milchakov** [comment]
@mgsergio PTAL

File: editor/feature_matcher.hpp [hunk 6]
```diff
@@ -5,16 +5,28 @@
...
-                                       vector<m2::PointD> const & geometry);
+                                       std::vector<m2::PointD> const & geometry);
 /// Returns value form [-1, 1]. Negative values are used as penalty, positive as score.
-/// |lhs| and |rhs| - triangulated polygons;
-double ScoreTriangulatedGeometries(vector<m2::PointD> const & lhs, vector<m2::PointD> const & rhs);
+/// |lhs| and |rhs| - triangulated polygons.
+/// Throws NotAPolygon exception when lhs or rhs is not convertible to a polygon.
+double ScoreTriangulatedGeometries(std::vector<m2::PointD> const & lhs,
+                                   std::vector<m2::PointD> const & rhs);
+/// Deprecated, use ScoreTriangulatedGeometries instead. In old versions duplicated points were removed
```
Comment (mgsergio) [inline editor/feature_matcher.hpp]: In older versions of editor the necessary geometry informations was missed form edits.xml file, so we cannot restore the original geometry of a particular feature and thus can't use the new algo that is dependant on a correct feature geometry to calculate scores.

Comment (mgsergio) [reply editor/feature_matcher.hpp]: @therearesomewhocallmetim, could you please review comments?

Comment (milchakov) [reply editor/feature_matcher.hpp]: готово

File: editor/feature_matcher.hpp [hunk 7]
```diff
@@ -5,16 +5,29 @@
...
 
-#include "std/vector.hpp"
+#include "base/exception.hpp"
+
+#include <vector>
 
 namespace matcher
 {
+DECLARE_EXCEPTION(NotAPolygonException, RootException);
 /// Returns closest to the latLon node from osm or empty node if none is close enough.
```
Comment (therearesomewhocallmetim) [inline editor/feature_matcher.hpp]: Returns a node from OSM closest to the latLon, or an empty node if none is close enough.

File: editor/feature_matcher.hpp [hunk 8]
```diff
@@ -5,16 +5,29 @@
...
 /// Returns a way from osm with similar geometry or empty node if can't find such way.
+/// Throws NotAPolygon exception when |geometry| is not convertible to a polygon.
 pugi::xml_node GetBestOsmWayOrRelation(pugi::xml_document const & osmResponse,
-                                       vector<m2::PointD> const & geometry);
+                                       std::vector<m2::PointD> const & geometry);
 /// Returns value form [-1, 1]. Negative values are used as penalty, positive as score.
-/// |lhs| and |rhs| - triangulated polygons;
-double ScoreTriangulatedGeometries(vector<m2::PointD> const & lhs, vector<m2::PointD> const & rhs);
+/// |lhs| and |rhs| - triangulated polygons.
+/// Throws NotAPolygon exception when lhs or rhs is not convertible to a polygon.
```
Comment (therearesomewhocallmetim) [inline editor/feature_matcher.hpp]: Throws NotAPolygon exception when either lhs or rhs is not convertible to a polygon.

File: editor/feature_matcher.hpp [hunk 9]
```diff
@@ -5,16 +5,29 @@
...
+double ScoreTriangulatedGeometries(std::vector<m2::PointD> const & lhs,
+                                   std::vector<m2::PointD> const & rhs);
+/// Deprecated, use ScoreTriangulatedGeometries instead. In older versions of editor the necessary
+/// geometry informations was missed form edits.xml file, so we cannot restore the original geometry
+/// of a particular feature and thus can't use the new algo that is dependant on a correct feature
+/// geometry to calculate scores.
+// TODO(a): To remove it when version 8.0.x will no longer be supported.
```
Comment (therearesomewhocallmetim) [inline editor/feature_matcher.hpp]: Deprecated, use ScoreTriangulatedGeometries instead. In the older versions of the editor, the edits.xml file didn't contain the necessary geometry information, so we cannot restore the original geometry of a particular feature and thus can't use the new algo that is dependent on correct feature geometry to calculate scores.
 TODO(a): To remove it when version 8.0.x is no longer supported.

File: editor/feature_matcher.hpp (hunk 7 above)
Comment (milchakov) [reply editor/feature_matcher.hpp]: готово

File: editor/feature_matcher.hpp (hunk 8 above)
Comment (milchakov) [reply editor/feature_matcher.hpp]: заменил

File: editor/feature_matcher.hpp (hunk 9 above)
Comment (milchakov) [reply editor/feature_matcher.hpp]: заменил

**Comment by milchakov** [comment]
@mgsergio PTAL&MP

**Pull request merged**